### ISSUE TITLE
Autocrop feature breaks DjVu support (possibly PDF also)

### ISSUE THREAD

**Issue by tigran123** [koreader/koreader#2205]
- KOReader version: v2015.11-549
- Device: Kobo Aura H2O, Kobo Aura HD, Kobo Touch and Kobo Mini (but reproducible on emulator as well)
#### Steps to reproduce

Open a DjVu file (for example this one:  A.S. Eddington, "The Nature of the Physical World", 1929 available here: http://libgen.io/book/index.php?md5=C76B75123D8A7700CDD4B1CB1BC05891) and set crop mode to "Auto". Start turning the pages and notice that you cannot consistently turn faster than about 1 page in 2 seconds. If you login via usbnet and run htop (from @NiLuJe 's wonderful Kobo-Stuff collection) you will notice that luajit process is taking 100% cpu and until it drops to 0% you cannot continue. Now exit Koreader and start it again. Notice that the open time increased from about 0.3 seconds to 15 seconds, i.e. slowed down FIFTY times. This is what I mean by "completely broken". DjVu format is famous for practically instant opening (regardless of document size --- even 3GB DjVu files open instantly) and extremely fast page decoding. The present state of Autocrop in Koreader completely destroys both.

Now, set crop mode to Manual (and select approx. the same rectangle as Auto would have chosen). Now you can turn at least 5 pages per second (even higher, but eInk screen can't keep up rendering them), that is 10 times faster. Also, opening time is reduced back to 0.3 seconds if you set Crop to Manual.

This is easily reproducible even on the emulator. Of course the modern desktops' Intel processors are much faster than the ARM processor used in Kobos, so you won't be able to noticeably slow down turning the pages, but by monitoring the cpu usage of luajit process you will see the same behaviour (with Mode set to Auto vs Manual).
#### Issue

I hope everyone agrees that if some function is slowed down by a factor of 50 it can be considered completely broken. So, that is what Autocrop is doing to DjVu support (probably to PDF also, but I haven't checked). 

I have tracked it down to the function `masterinfo_new_source_page_init()` called from `libk2pdfopt/koptcrop.c:k2pdfopt_crop_bmp()` and putting instrumentation in it I have further narrowed it down to the two calls it makes:

```
1. bmp_adjust_contrast():   0.03 seconds
2. bmp_convert_to_greyscale(src):   0.05 seconds
```

The timings are for emulator --- on Kobo it would be about 30 times slower, so 0.08 *30 = 2.4 seconds --- so this explains the two seconds delay introduced for each page. So, this "initialisation step" takes 9 times longer than the actual step of trimming the margins (`bmpregion_trim_margins():   0.010648 seconds`)!

Now, the first thing that comes to mind is to comment out these two calls. Unfortunately, the call to `bmp_adjust_contrast()` is needed for some reason (autocrop becomes not very reliable without it). However, the call to `bmp_convert_to_greyscale(src)` can be safely commented out and this hack sort of solves the issue, i.e. after this change:

```
diff --git a/k2pdfoptlib/k2master.c b/k2pdfoptlib/k2master.c
index fdb3136..b826ea8 100644
--- a/k2pdfoptlib/k2master.c
+++ b/k2pdfoptlib/k2master.c
@@ -231,8 +231,10 @@ printf("11\n");
 printf("22\n");
 #endif
     /* Convert source back to gray scale if not using color output */
+#if 0
     if (!k2settings_need_color_permanently(k2settings))
         bmp_convert_to_greyscale(src);
+#endif
     region->dpi = k2settings->src_dpi;
     region->r1 = 0;
     region->r2 = srcgrey->height-1;
```

you can now turn about 2-3 pages per second and the opening time is reduced down to about 1.5-2 seconds. This is much better than one page in two seconds and 15 seconds opening time, but still not as good as 6 pages per second and 0.3 seconds opening time if the Crop mode is set to Manual.

There seems to be nothing DjVu-specific about this problem, so it probably affects PDF in the same way also. Now, on this planet, properly typeset books come in just two formats: PDF and DjVu (plus the legacy "printed on paper" ones), so the fact that this breakage was unnoticed for such a long time (probably since the upgrade of k2pdfopt to v2.34 in May 2016) confirms what I suspected for a long time, namely that nobody reads books anymore nowadays --- only those things called "ebooks": you know, the linear streams of characters which are turned into glyphs on the fly by the software on the device in real time, instead of letting professional master typographers do their mysterious art "offline".  I consider Koreader to be primarily a Book reading software (not eBooks) and so I consider this breakage to be catastrophic, i.e., sadly, terminating Koreader's life if it is not fixed properly. (For my own purposes of using Koreader for reading the above hack is sufficient, but I realise that this is not a proper solution and may break colour screens like Android tablets etc.)

Initially this problem was mentioned in #2130 but I decided to separate all the useful information into this issue specifically dedicated to it.

**Comment by houqp** [comment]
The fact that it takes so much computing power to do the greyscale comparison doesn't make sense to me... We should take a deep dive into that function to see what's going wrong.

PS: Actually most of the books I read are in PDF :P But I am also a slower reader...

**Comment by tigran123** [comment]
@houqp Thank you, but I have two questions:

1. What is the point of optimizing that greyscale conversion if it is not needed in this case at all? Why not comment it out altogether as I have done in the above patch?

2. I didn't take a deep dive into that function because I think it would be much more sensible to NOT use k2pdfopt module for function which has nothing to do with enabling Reflow. Namely, a standalone "trim margins" function should be written that will be at best about 100,000 times faster than the current monstrosity we have and at worst at least 10,000 times faster. Even if it is just 2,000 times faster I would be happy. So my question is: don't you think this would be much more sensible than optimizing greyscale copy (or some other optimization) which at most would give you a speedup factor of 3x or 5x?

**Comment by chrox** [comment]
Interesting. Did you try reverting libk2pdfopt to `5aa0165519ec4d6dc74e12aa231f184be85e8a9b` to see if it is introduced in k2pdfopt v2.34? And it will be nice to also test if it affects the performance of reflowing.

I will reimplement the autocrop function with `leptonica` and hopefully it will have much better performance.

**Comment by tigran123** [comment]
Thank you, @chrox  --- I knew you would suggest a proper solution and I am so glad you did! :)

No, to be honest, I did not test with the earlier k2pdfopt than v2.34 --- it was only a suspicion that the slowdown is caused by that upgrade, but not 100% proven.

But let me emphasize that the actual _fact_ of slowdown (by an enormous factor of 30x for page turns and 50x for djvu file opening) is confirmed beyond any shadow doubt on Kobo and even reproduced on the emulator. So this part is absolutely certain.

As for reflow --- looking at the way it behaves, I think it internally does auto-cropping as well, so it should affect its behaviour too (and slow it down, just like without reflow). So, your solution should improve the reflow performance as well.

Oh btw, I also noticed this --- when you return from reflow (i.e. set it to Off) then the crop is reset to "page fit" and the scroll is reset to Off, even if both these two settings were set to, say, "Manual" and "ON" respectively prior to entering Reflow mode. I think it would be more correct to restore them to the previous setting instead of resetting, but this is a minor cosmetic issue, of course. Don't worry about it.

**Comment by chrox** [comment]
> when you return from reflow (i.e. set it to Off) then the zoom mode is reset to "page fit" and the scroll is reset to Off, even if both these two settings were set to, say, "Crop: Auto" and "ON" respectively prior to entering Reflow mode. 

This sounds like a new bug and I cannot reproduce it on my Kindle.

**Comment by houqp** [comment]
> What is the point of optimizing that greyscale conversion if it is not needed in this case at all? Why not comment it out altogether as I have done in the above patch?

As you mentioned earlier, it might break devices with color screens.

> I didn't take a deep dive into that function because I think it would be much more sensible to NOT use k2pdfopt module for function which has nothing to do with enabling Reflow. Namely, a standalone "trim margins" function should be written that will be at best about 100,000 times faster than the current monstrosity we have and at worst at least 10,000 times faster. Even if it is just 2,000 times faster I would be happy. So my question is: don't you think this would be much more sensible than optimizing greyscale copy (or some other optimization) which at most would give you a speedup factor of 3x or 5x?

I totally agree. This will be the ideal solution. But in case it would take too long to reimplement this feature, fixing it k2pdfopt would not be a bad choice considering it feels like a regression in a recent update. I would not suggest this if this feature has been broken since day one.


> > when you return from reflow (i.e. set it to Off) then the zoom mode is reset to "page fit" and the scroll is reset to Off, even if both these two settings were set to, say, "Crop: Auto" and "ON" respectively prior to entering Reflow mode.
>
> This sounds like a new bug and I cannot reproduce it on my Kindle.

Not able to reproduce on my side either, @tigran123 is this happening to all your books?

**Comment by tigran123** [comment]
Please ignore the second issue --- the minor cosmetic one I mentioned about returning from reflow losing the previous zoom/scroll settings. This must have been happening with the old Koreader versions as the current versions don't exhibit this behaviour --- the previous setting is correctly restored, I have just tested it.

So, let's concentrate on the main issue of the huge slowdown of page turning and opening (not initial, but subsequent to setting Crop to Auto) of djvu files.

**Comment by chrox** [comment]
I tested the reflowing performance of the latest nightly with a benchmark of the last stable release on my Kindle Voyage.

Reflowing duration is recorded for each page and plotted versus page number:

![reflowing_benchmark](https://cloud.githubusercontent.com/assets/751535/17555897/ed403f58-5f43-11e6-9319-635342816532.png)

For the same 50 pages of a typical PDF document the average duration increases 23% in the latest nightly build compared to the last stable release. This is still acceptable considering that k2pdfopt adds a lot of features enhancements in the last whole year.

I'll also benchmark the performance of the `leptonica` autocrop with the `k2pdfopt` version in this way.

**Comment by tigran123** [comment]
Thank you, very interesting graph. But we still have the fact of 30 (yes THIRTY TIMES) performance degradation when turning DjVu pages and 50 (yes, FIFTY TIMES) performance degradation when opening djvu file (after autocrop set to Auto).

**Comment by chrox** [comment]
Yes, I'm providing this graph to exemplify the right scientific methodology. When we talk about performance degradation we must specify what to compare. We can only compare the performance of the same function to measure the degradation but you claimed "autocrop" is 30x to 50x slower than "manual crop" about which I don't doubt, to be performance degradation. I would say it's conceptually wrong like measuring color degradation between apple and orange. 

Auto-cropping by its nature will be much slower than manual crop. In short, auto-cropping will at least read all the pixels of a page to calculate the average background color and then trim the margin. Note that using hard-coded 128/192 brightness as background threshold won't have good results for a large part of scanned book even though it will save the process of background calculation. So comparing the time of "manual crop" (null calculation) with "auto crop" (image processing calculation) is irrational and has no meaning.

The right methodology is comparing the performance of "autocrop" function in the stable release with the nightly release just like the above. And you did identify a performance hotspot `bmp_convert_to_greyscale` in the k2pdfopt library and [the hotspot is new code in v2.34](https://github.com/koreader/libk2pdfopt/blame/e8fab06fdcb321b06ce46dba4b4a70847b98ed56/k2pdfoptlib/k2master.c#L234-L235).

**Comment by tigran123** [comment]
Well, yes, I agree that one shouldn't compare auto crop with manual crop as such. But what I meant was that in the past the difference between the two was too slight to be perceived, whereas now it became enormous and annoying. Only in this specific sense I referred to manual crop (as an indicator of the "old performance of autocrop"), but not in the absolute sense which would be, as you rightly say, comparing apples and oranges.

**Comment by chrox** [comment]
And here is the number:

![autobbox_benchmark](https://cloud.githubusercontent.com/assets/751535/17593099/5e2f9140-6017-11e6-851a-81fbe822ecdb.png)

Latest nightly build is 69% slower than the last stable release on auto-cropping. And the degradation will be eliminated by #2217.

But the problem you brought up still exists as you can see in the above graph, there are spikes that takes unbearably long time to finish. It needs further patches to optimize to really close this issue. Thanks for your report and investigation and #2217 is just commenting out:
``` c
if (!k2settings_need_color_permanently(k2settings))
         bmp_convert_to_greyscale(src);
```
as its useless to convert to grayscale because bmps to be reflowed in koreader are already grayscale.

**Comment by tigran123** [comment]
Thank you, @chrox I really like your very thorough and scientific approach. It guarantees clarity and prevents ambiguity or surprises! So, my intuitive guess that that conversion to grayscale can be thrown away was correct, after all :)

**Comment by NiLuJe** [comment]
Fixed by #2217, right?
